sparksee/src/main/java/com/yahoo/ycsb/db/sparksee/SparkseeClient.java

```diff
@@ -0,0 +1,359 @@
+import com.yahoo.ycsb.generator.graph.Edge;
+import com.yahoo.ycsb.generator.graph.Node;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Sparksee client for the YCSB benchmark.
+ */
+public class SparkseeClient extends DB {
+
+  static final String SPARKSEE_DATABASE_PATH_PROPERTY = "sparksee.path";
+  private static final String SPARKSEE_DATABASE_PATH_DEFAULT = "sparkseeDB.gdb";
+  private static final String SPARKSEE_LOG_LEVEL_PROPERTY = "sparksee.logLevel";
+  private static final String USE_INDEX_PROPERTY = "sparksee.index";
+  private static final String USE_INDEX_DEFAULT = "false";
+  private static final String SPARKSEE_LOG_LEVEL_DEFAULT = "Off";
+  private static final Object INIT_LOCK = new Object();
+  private static final AtomicInteger INIT_COUNT = new AtomicInteger();
+
+  private static boolean initialised = false;
+  private static Sparksee sparksee;
+  private static Database database;
+  private static Integer nodeIdAttribute = null;
+  private static Integer edgeIdAttribute = null;
+  private boolean useIndex;
+
+  @Override
+  public void init() throws DBException {
+    super.init();
+
+    INIT_COUNT.incrementAndGet();
+
+    synchronized (INIT_LOCK) {
+      if (!initialised) {
+        Properties properties = getProperties();
+
+        String path = properties.getProperty(SPARKSEE_DATABASE_PATH_PROPERTY, SPARKSEE_DATABASE_PATH_DEFAULT);
+        String logLevel = properties.getProperty(SPARKSEE_LOG_LEVEL_PROPERTY, SPARKSEE_LOG_LEVEL_DEFAULT);
+        useIndex = Boolean.parseBoolean(properties.getProperty(USE_INDEX_PROPERTY, USE_INDEX_DEFAULT));
+
+        SparkseeConfig sparkseeConfig = new SparkseeConfig();
+        setLogLevel(sparkseeConfig, logLevel);
+
+        sparksee = new Sparksee(sparkseeConfig);
+
+        try {
+          if (new File(path).exists()) {
+            database = sparksee.open(path, false);
+          } else {
+            database = sparksee.create(path, "SparkseeDB");
+          }
+        } catch (FileNotFoundException e) {
+          e.printStackTrace();
+        }
+
+        try (Session session = database.newSession()) {
+          Graph graph = session.getGraph();
+
+          nodeIdAttribute = getAttribute(graph, getNodeType(graph), "sparksee.nodeId");
+          edgeIdAttribute = getAttribute(graph, getEdgeType(graph), "sparksee.edgeId");
+
+          if (useIndex) {
+            try {
+              graph.indexAttribute(nodeIdAttribute, AttributeKind.Indexed);
+              graph.indexAttribute(edgeIdAttribute, AttributeKind.Indexed);
+            } catch (RuntimeException e) {
+              e.printStackTrace();
+            }
+          }
+        }
+
+        initialised = true;
+      }
+    }
+  }
+
+  @Override
+  public void cleanup() throws DBException {
+    if (INIT_COUNT.decrementAndGet() == 0) {
+      database.close();
+      sparksee.close();
+      initialised = false;
```

Comment:
this needs to be protected by a lock on INIT_COUNT similar to how the init method is guarded.